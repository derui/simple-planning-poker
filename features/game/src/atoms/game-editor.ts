import { UseLoginUser } from "@spp/feature-login";
import { GameRepository, User } from "@spp/shared-domain";
import { EditGameUseCase } from "@spp/shared-use-case";
import { useAtom, useSetAtom } from "jotai";
import { GameDto, toGameDto } from "./dto.js";
import { gamesAtom, GameStatus, gameStatusAtom, selectedGameAtom } from "./game-atom.js";

/**
 * Hook definition to edit game
 */
export type UseGameEditor = () => {
  loading: boolean;

  /**
   * current selected game.
   */
  game?: GameDto;

  /**
   * edit current selected game.
   */
  edit: (name: string, points: string) => void;
};

/**
 * Create hook implementation of `UseGameEditor`
 */
export const createUseGameEditor = function createUseGameEditor({
  gameRepository,
  useLoginUser,
  editGameUseCase,
}: {
  gameRepository: GameRepository.T;
  useLoginUser: UseLoginUser;
  editGameUseCase: EditGameUseCase;
}): UseGameEditor {
  return () => {
    const [gameStatus, setGameStatus] = useAtom(gameStatusAtom);
    const { userId } = useLoginUser();
    const [game, setGame] = useAtom(selectedGameAtom);
    const setGames = useSetAtom(gamesAtom);
    const loading = gameStatus == GameStatus.Editing;

    return {
      loading,
      game: game ? toGameDto(game) : undefined,

      edit: (name: string, points: string): void => {
        if (!game || !userId || loading) {
          return;
        }

        const _do = async () => {
          try {
            const ret = await editGameUseCase({
              gameId: game.id,
              name,
              points: points.split(",").map((v) => parseInt(v)),
              ownedBy: userId,
            });

            switch (ret.kind) {
              case "success":
                setGame(ret.game);
                setGames(await gameRepository.listUserCreated(User.createId(userId)));
                break;
              default:
                break;
            }
          } catch (err) {
            console.error(err);
          } finally {
            setGameStatus(GameStatus.Selected);
          }
        };

        setGameStatus(GameStatus.Editing);
        _do();
      },
    };
  };
};
