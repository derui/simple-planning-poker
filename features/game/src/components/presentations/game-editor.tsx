import { zodResolver } from "@hookform/resolvers/zod";
import { Variant } from "@spp/shared-color-variant";
import { ApplicablePoints } from "@spp/shared-domain";
import { Input } from "@spp/ui-input";
import { Loader } from "@spp/ui-loader";
import { SubmitHandler, useForm } from "react-hook-form";
import * as z from "zod";
import { CreateGameError } from "../../atoms/game-atom.js";
import * as styles from "./game-editor.css.js";

interface Props {
  defaultName?: string;

  defaultPoints?: string;

  /**
   * do submit after validate is succeeded
   */
  onSubmit?: (name: string, points: string) => void;

  /**
   * validate input. return errors if exists. This function should return empty array if all argument is valid.
   */
  onValidate?: (name: string, points: string) => CreateGameError[];

  onCancel?: () => void;

  loading?: boolean;
}

interface Input {
  name: string;
  points: string;
}

const schema = z.object({
  name: z.string().min(1),
  points: z
    .string()
    .min(1)
    .refine((v) => ApplicablePoints.parse(v), { message: "Invalid points" }),
});

type Scheme = z.infer<typeof schema>;

// eslint-disable-next-line func-style
export function GameEditor(props: Props): JSX.Element {
  const { defaultName, defaultPoints, onSubmit, loading = false, onCancel, onValidate } = props;

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<Input>({
    resolver: zodResolver(
      schema.refine(
        (args): args is Scheme => {
          const { name, points } = args;
          const errors = onValidate?.(name, points) ?? [];

          if (errors.length > 0) {
            return errors.every((e) => e != "NameConflicted");
          } else {
            return true;
          }
        },
        { path: ["name"], message: "Given name is conflicted" }
      )
    ),
  });

  const handleWrappedSusbmit: SubmitHandler<Input> = (data) => {
    onSubmit?.(data.name, data.points);
  };

  return (
    <form className={styles.root} onSubmit={handleSubmit(handleWrappedSusbmit)}>
      <div className={styles.input.row}>
        <label htmlFor="name" className={styles.input.label}>
          Name
        </label>
        <Input
          type="text"
          id="name"
          placeholder="e.g. A sprint"
          className={styles.input.input}
          defaultValue={defaultName}
          tabIndex={0}
          disabled={loading}
          {...register("name")}
        />
        {errors?.name?.message ? <p className={styles.input.error}>{errors.name.message}</p> : undefined}
      </div>
      <div className={styles.input.row}>
        <label htmlFor="points" className={styles.input.label}>
          Points
        </label>
        <Input
          type="text"
          id="points"
          className={styles.input.input}
          defaultValue={defaultPoints}
          tabIndex={0}
          disabled={loading}
          {...register("points")}
        />
        {errors?.points?.message ? <p className={styles.input.error}>{errors.points.message}</p> : undefined}
      </div>
      <div className={styles.footer.root}>
        <button type="button" className={loading ? styles.cancelDisabled : styles.cancel} onClick={onCancel}>
          Cancel
        </button>
        <button className={loading ? styles.submitLoading : styles.footer.submit} type="submit" aria-busy={loading}>
          <Loader size="s" shown={loading} variant={Variant.emerald} />
          {loading ? "" : "Submit"}
        </button>
      </div>
    </form>
  );
}
