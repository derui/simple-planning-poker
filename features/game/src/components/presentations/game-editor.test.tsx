import { cleanup, render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import sinon from "sinon";
import { afterEach, expect, test } from "vitest";
import { GameEditor } from "./game-editor.js";

afterEach(cleanup);

test("should be able to render", async () => {
  const ret = render(<GameEditor />);

  expect(ret.container).toMatchSnapshot();
});

test("call callback after submit", async () => {
  // Arrange
  const callback = sinon.fake();

  render(<GameEditor onSubmit={callback} />);

  // Act
  await userEvent.type(screen.getByLabelText("Name"), "test");
  await userEvent.type(screen.getByLabelText("Points"), "1,2,3");
  await userEvent.click(screen.getByText("Submit"));

  // Assert
  expect(callback.calledOnceWith("test", "1,2,3")).toBeTruthy();
});

test("show initial value", async () => {
  // Arrange
  render(<GameEditor defaultName="name" defaultPoints="1,2,3" />);

  // Act

  // Assert
  expect(screen.getByLabelText<HTMLInputElement>("Name").value).toEqual("name");
  expect(screen.getByLabelText<HTMLInputElement>("Points").value).toEqual("1,2,3");
});

test("call cancel callback", async () => {
  // Arrange
  const fake = sinon.fake();
  render(<GameEditor defaultName="name" defaultPoints="1,2,3" onCancel={fake} />);

  // Act
  await userEvent.click(screen.getByText("Cancel"));

  // Assert
  expect(fake.calledOnce).toBe(true);
});

test("do not callback if validation returns specific error", async () => {
  // Arrange
  const submit = sinon.fake();
  render(<GameEditor defaultName="name" defaultPoints="1,2,3" onSubmit={submit} errors={["NameConflicted"]} />);

  // Act

  // Assert
  expect(screen.getByLabelText<HTMLInputElement>("Name").value).toEqual("name");
  expect(screen.getByLabelText<HTMLInputElement>("Points").value).toEqual("1,2,3");
  await userEvent.click(screen.getByText("Submit"));

  expect(submit.calledOnce).toBe(false);
});

test("call submit if custom callback returns no errors", async () => {
  // Arrange
  const submit = sinon.fake();
  render(<GameEditor defaultName="name" defaultPoints="1,2,3" onSubmit={submit} errors={[]} />);

  // Act

  // Assert
  expect(screen.getByLabelText<HTMLInputElement>("Name").value).toEqual("name");
  expect(screen.getByLabelText<HTMLInputElement>("Points").value).toEqual("1,2,3");
  await userEvent.click(screen.getByText("Submit"));

  expect(submit.calledOnce).toBe(true);
});
