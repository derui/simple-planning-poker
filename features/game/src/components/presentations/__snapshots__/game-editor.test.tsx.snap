// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should be able to render 1`] = `
<div>
  <form
    class="game-editor_root__nauw780"
  >
    <div
      class="game-editor_input_row__nauw782"
    >
      <label
        class="game-editor_input_label__nauw781"
        for="name"
      >
        Name
      </label>
      <input
        class="style_input__xo256r0 style_transition_border__192bcwg42"
        id="name"
        name="name"
        placeholder="e.g. A sprint"
        tabindex="0"
        type="text"
        value=""
      />
    </div>
    <div
      class="game-editor_input_row__nauw782"
    >
      <label
        class="game-editor_input_label__nauw781"
        for="points"
      >
        Points
      </label>
      <input
        class="style_input__xo256r0 style_transition_border__192bcwg42"
        id="points"
        name="points"
        tabindex="0"
        type="text"
        value=""
      />
    </div>
    <div
      class="game-editor_footer_root__nauw788"
    >
      <button
        class="game-editor_cancel__nauw786 style_enabledButtonVariants_gray__92u52zo style_enabledBase__92u52z3 style_base__92u52z0 style_transition_all__192bcwg41"
        type="button"
      >
        Cancel
      </button>
      <button
        aria-busy="false"
        class="game-editor_footer_submit__nauw789 style_enabledButtonVariants_emerald__92u52zr style_enabledBase__92u52z3 style_base__92u52z0 style_transition_all__192bcwg41 game-editor_submitBase__nauw785"
        type="submit"
      >
        <span
          class="style_hidden__1yu8yli1"
          data-shown="false"
          role="alert"
        >
          <span
            class="style_spinLoader_s__1yu8yli6 style_spinLoaderBase__1yu8yli5 style_animation_spin__192bcwg3z"
          >
            <span
              class="loader2_small_emerald__120am7c3 style_iconSmall__p90ga81 style_iconBase__p90ga80"
            />
          </span>
        </span>
        Submit
      </button>
    </div>
  </form>
</div>
`;
